mloop/controllers.py

```diff
@@ -53,8 +54,27 @@ def create_controller(interface,
     elif controller_type=='random':
         controller = RandomController(interface, **controller_config_dict)
     else:
-        log.error('Unknown controller type:' + repr(controller_type))
-        raise ValueError
+        # If `controller_type` doesn't match any of the built-in controllers,
+        # looks for an external controller with matching name.
+        # The `controller_type` should be "module_name:controller_name".
+        parts = controller_type.split(":")
+        if (len(parts)==2) and all([len(part)>0 for part in parts]):
+            try:
+                module = import_module(parts[0])
+                constructor = getattr(module, parts[1])
+            except ModuleNotFoundError:
+                log.error(f"Unknown module: {parts[0]}")
+                raise ValueError
+            except AttributeError:
+                log.error(f"Unknown controller type: {parts[0]}.{parts[1]}")
+                raise ValueError
+            except:
+                log.error('Unknown controller type:' + repr(controller_type))
+                raise ValueError
+            controller = constructor(interface, **controller_config_dict)
```

Comment:
Do you know what form the config dict takes, in particular whether the values are parsed or raw strings? (E.g. if `test_point_count = 100` is in the file, will the dict have `{"test_point_count": 100}` or `{"test_point_count": "100"}`?)

Reply:
The way M-LOOP parses the dict seems to be to take the entire configuration file, turn it into a single `tdict_string`, and then run `eval` on `dict({tdict_string})`, so I'd expect that `test_point_count = 100` becomes `{"test_point_count": 100}`. I added some print statements to be sure, and I think that's how it works.

(I had to test if this way of parsing the files meant that I could run random code from the configuration file, and it seems that writing something like `) and __import__("sys").exit (0` at the end of file does make M-LOOP exit.)

Reply:
😆 nice find!

I think that means we should add a factory function (`qctrl_controller` or somesuch) to `qctrlmloop` that accepts regular old primitive arguments (instead of things like `Learner` and `Qctrl` objects), which can convert those into the relevant `Learner` and `Qctrl` objects before creating the actual `QctrlController`. But that's something we can figure out in `qctrlmloop`.

I'm happy with this as-is, so just mark it as ready for review when you're ready and I'll merge it.

Reply:
Ok, sounds good! I'll mark as ready for review